CT integration and Generation of thresholded mesh from CT

chinmaychinara91 opened the pull request:
# CT integration

This PR adds: 
* New node type `'volct'` for CT data, icon and tooltip (It requires `brainstorm.jar` from [here](https://github.com/rcassani/bst-java/blob/add-ct/brainstorm/dist/brainstorm.jar))

![image](https://github.com/brainstorm-tools/brainstorm3/assets/8238803/9fc3c7ef-dfc9-4781-8cfa-27ad76aca2fa)

* Updated popup menu in the database tree for `'volct'`
    * MRI-exclusive actions are removed
    * Do not allow to be Subject's default MRI
    * Do not allow editing MRI form CT

![image](https://github.com/brainstorm-tools/brainstorm3/assets/8238803/d91ddefa-7130-4b40-9d2a-fffdacee42ca)

* Example of updating a process to handle MRI and CT: `process_mri_deface.m`
![image](https://github.com/brainstorm-tools/brainstorm3/assets/8238803/c1225a35-3546-4639-b62e-909d4818713d)


# Generation of thresholded mesh from CT
**[THESE ARE USED AS TEST DATA BELOW](https://drive.google.com/drive/folders/1TI8wANAb09erddYP52bXyXo5xuajLCNf?usp=sharing)**

**NOTE: This is just a CT specific feature**

1. Assuming we have a CT by following [Import CT](https://neuroimage.usc.edu/brainstorm/Tutorials/ImportAnatomy#Import_CT)


2. Right-click on the **CT → MRI segmentation → Generate mesh from CT** . 
<img src="https://github.com/brainstorm-tools/brainstorm3/assets/21283733/d720129c-4bc0-44b7-95a4-e65040b4778b"  width="500" />


3. This will bring the **Generate isosurface** window. This window shows 3 things in the Hounsfield Unit, the Background level, the White level and the Set isoValue fields. The first and second values are calculated automatically from the histogram of the CT. This is to help the user to decide on what to set the isoValue for getting a good thresholded mesh. Another way is to refer to this https://en.wikipedia.org/wiki/Hounsfield_scale. The best way, however is to just open the CT in the MRI viewer browse through the electrode intensities and do trial and error to get the best value for a clean mesh. In this case, setting the value to **2400** gave a good result. Set the **isoValue** to **2400** and press **OK.** 
<img src="https://github.com/brainstorm-tools/brainstorm3/assets/21283733/516247ef-9780-4ab1-8279-f67b7bf94a98"  width="200" />


4. This will generate a isosurface which looks something like under. **THIS WILL HELP IN DOING SEEG ELECTRODE CONTACT LOCALIZATION**.
<img src="https://github.com/brainstorm-tools/brainstorm3/assets/21283733/54e48822-0a24-4ecf-a88a-1f292b2f7e64"  width="500" />

rcassani commented on toolbox/anatomy/tess_isosurface.m:
```diff
@@ -0,0 +1,156 @@
+    return
+end
+
+%% ===== ASK PARAMETERS =====
+% Ask user to set the parameters if they are not set
+if (nargin < 2) || isempty(isoValue)
+    res = java_dialog('input', {'<HTML>Background level (HU):<BR>(guessed from MRI histogram)', ...
+                                '<HTML>White level (HU):<BR>(guessed from MRI histogram)', ...
+                                '<HTML>Set isoValue (HU):'}, ...
+                                'Generate isosurface', ...
+                                [], ...
+                                {num2str(sMri.Histogram.bgLevel), num2str(sMri.Histogram.whiteLevel), num2str(1900)});
```
The first and second inputs are not used for now, right?
If they are not going to be used in the future, we should just print them instead of being a editable fields.

chinmaychinara91 replied:
That is correct. Let me make these changes.

rcassani commented on toolbox/anatomy/tess_isosurface.m:
```diff
@@ -0,0 +1,156 @@
+    return
+end
+
+%% ===== ASK PARAMETERS =====
+% Ask user to set the parameters if they are not set
+if (nargin < 2) || isempty(isoValue)
+    res = java_dialog('input', {'<HTML>Background level (HU):<BR>(guessed from MRI histogram)', ...
+                                '<HTML>White level (HU):<BR>(guessed from MRI histogram)', ...
+                                '<HTML>Set isoValue (HU):'}, ...
+                                'Generate isosurface', ...
+                                [], ...
+                                {num2str(sMri.Histogram.bgLevel), num2str(sMri.Histogram.whiteLevel), num2str(1900)});
```
I'm not familiar with CT data. 
Will this guess number (1900) change from CT to CT?
If so, we can we propose a better guess based on the CT data?

chinmaychinara91 replied:
That was my thought as well. But from all the CT data I have encountered till now that seems to be a good number based on manual inspection as mentioned in the PR description above. For now let's keep it to 1900. But if we find any new ways to better initialize this value in the future will update it.

chinmaychinara91 replied:
Check https://github.com/brainstorm-tools/brainstorm3/pull/684/commits/ea8777bffd01536a3559a3860c444237a1433900. Found a better way using `sMRI.Histogram.maxIntensity`.  Added a parameter check for that as well since anything above `sMRI.Histogram.maxIntensity` retunrs empty and there is nothing to compute.

rcassani commented on toolbox/anatomy/tess_isosurface.m:
```diff
@@ -0,0 +1,156 @@
+    % Get new value isoValue
+    isoValue = round(str2double(res{3}));
+else
+    isoValue = sMri.Histogram.whiteLevel;
```
Potential error, this overrides the input argument given to the function

chinmaychinara91 replied:
Yes thanks for catching this.

rcassani reviewed (changes requested):
Overall the code looks good to me.
Few things to address.

rcassani commented:
@chinmaychinara91, check the review above.

There are few commits that I want to add on integration of CT now in Brainstorm, and I think this PR is a good place to put them so we take it as common place for further development in the sEEG contact location. What do you think?  

The two small modifications on `mri_coregister` added to this branch, as part of those modifications.

chinmaychinara91 commented:
@rcassani:

The `mri_coregister` changes look good. Others I will update.

chinmaychinara91 commented:
@rcassani updated with the requested changes.

chinmaychinara91 commented:
just found another issue. let me update it.

chinmaychinara91 commented:
@rcassani done from my end.

rcassani commented:
I have updated the integration of CT data into Brainstorm: 
 * By creating its own node type we can handle better its behaviour, display and menus
 * The MRI menus that do not apply to CT have been removed
     * @tmedani, What do you think of the updated menu for CT? 
       Do we want to keep the 'MNI normalization' option for CT data? 
     
@chinmaychinara91, please test that I did not break anything with these changes.
To test the new node type the file:
1. The file `brainstorm3\java\brainstorm.jar` must be replaced with the one in this one: https://github.com/rcassani/bst-java/blob/add-ct/brainstorm/dist/brainstorm.jar
2. Matlab needs to be restarted as well

---
The title and description of this PR have been updated to match the contents

tmedani commented:
> I have updated the integration of CT data into Brainstorm:
> 
> * By creating its own node type we can handle better its behaviour, display and menus
> * The MRI menus that do not apply to CT have been removed
>   
>   * @tmedani, What do you think of the updated menu for CT?

I need to check the menu first for a complete feedback, 

>     Do we want to keep the 'MNI normalization' option for CT data?

I think we can keep it since it can be used in some other scenario, and it is always useful.

> The MRI 3D slices or,

Let's do it on the 3D MRI viewer, it has more significance.

chinmaychinara91 commented on toolbox/anatomy/tess_isosurface.m:
```diff
@@ -0,0 +1,155 @@
+function [MeshFile, iSurface] = tess_isosurface(iSubject, isoValue, Comment)
+% TESS_ISOSURFACE: Reconstruct a thresholded surface mesh from a CT
+%
+% USAGE:  [MeshFile, iSurface] = tess_isosurface(iSubject, isoValue=1900, Comment)
+%         [MeshFile, iSurface] = tess_isosurface(MriFile,  isoValue=1900, Comment)
+%         [Vertices, Faces]    = tess_isosurface(sMri,     isoValue=1900)
```
doing it now

chinmaychinara91 commented:
Also we both reviewed the changes they look great and work perfect.

chinmaychinara91 commented on toolbox/tree/tree_callbacks.m:
```diff
@@ -176,6 +176,19 @@
                 end
+
+            % ===== VOLUME CT =====
+            case 'volct'
```
can we maybe merge `case 'volct'` with `case 'volatlas'` to `case {'volatlas', 'volct'}` as they are doing the same thing essentially ?

rcassani replied:
Sounds good. Done at [8810927](https://github.com/brainstorm-tools/brainstorm3/pull/684/commits/8810927dcdea4f047b1f5c49ee4e7cf46e2d664a)

chinmaychinara91 commented:
all things look good @rcassani. Just wrt https://github.com/brainstorm-tools/brainstorm3/pull/684#discussion_r1501248906, as you suggested and as @tmedani agreed, we can add loading of just the MRI 3D slices along with the new mesh so that we get a good reference for the contacts.

chinmaychinara91 commented:
> all things look good @rcassani. Just wrt [#684 (comment)](https://github.com/brainstorm-tools/brainstorm3/pull/684#discussion_r1501248906), as you suggested and as @tmedani agreed, we can add loading of just the MRI 3D slices along with the new mesh so that we get a good reference for the contacts.

Updated and added this.

rcassani commented:
Great, I'm merging this PR!
@chinmaychinara91, please update the "Generation of thresholded mesh from CT" in the PR description to match the final version
